### PR TITLE
Don't ignore 4XX errors, add BadRequest

### PR DESCRIPTION
**Pull request by pbrisbin**
I was very puzzled to find that unknown 4XX statuses are just silently
accepted. Additionally, BadRequest is something we use frequently in our API
for missing parameters and the like, so this adds explicit support for that as
well.

### CODE REVIEW

File: lib/json_api_client/middleware/status.rb [hunk 1]
```diff
@@ -29,7 +31,7 @@ def handle_status(code, env)
         when 409
           raise Errors::Conflict, env
         when 400..499
-          # some other error
+          raise Errors::UnexpectedStatus.new(code, env[:url])
```
Comment (adisos) [inline lib/json_api_client/middleware/status.rb]: I think we should pass `env` variable also

**Comment by Fivell** [comment]
there should not be exception for client side errors.
See https://github.com/chingor13/json_api_client/pull/243#pullrequestreview-25112466

File: lib/json_api_client/middleware/status.rb (hunk 1 above)
Comment (adisos) [inline lib/json_api_client/middleware/status.rb]: And I'd better use `Errors:: ClientError` exception.

**Comment by pbrisbin** [comment]
I apologize, was working a little quickly on Friday and I was not detailed, specific, or even accurate in my PR description. Let me try again...

My actual use-case is returning 4XX (I think 422) on missing or invalid parameters (`filter`) on `GET` requests. Because this general middleware's behavior is so coupled to the far-away behavior of `#save` (non-`GET`), it's not behaving correctly for this case and I didn't understand the reasons why.

FWIW, had I been aware of those previous discussions, I might've suggested the middleware do raise exceptions, since that makes sense at that level of abstraction. but then `#save` could just rescue, convert any response errors (stored as `ex.errors`) to model errors, and return its own `false`.

Alternatively, would it be acceptable to raise a `ClientError` on 4XX if the method is `GET`? That would be slightly messy in the interest in backwards-compatibility, but would get me the behavior I'm looking for.

Also, if the addition of `BadRequest` is not controversial, I could pull that into its own PR so the discussions here don't hold up that aspect.

File: lib/json_api_client/middleware/status.rb (hunk 1 above)
Comment (pbrisbin) [inline lib/json_api_client/middleware/status.rb]: That sounds right to me. Once the discussion with @Fivell reaches a conclusion, I'll update to do the more correct thing here.

**Comment by Fivell** [comment]
@pbrisbin , "bang-methods" in this case sounds suitable 👍

**Comment by Fivell** [comment]
@pbrisbin , also moving  BadRequest to its own PR is also good idea, looks like it is related to  #216

**Comment by pbrisbin** [comment]
Sorry about the noise here. I'm going to hold off on proposing any of these changes for now (even the `BadRequest` one). After getting the hang of using `ResultSet#errors`, the problems that lead to this PR aren't present.

**Comment by ngouy** [comment]
Yea ! That was exactly what I was looking for, thank you ! Any news about when it will be pushed ? @adisos

**Comment by ngouy** [comment]
THank you @adisos , great thing. This is what I was looking for ! Any news about when it will be pushed ?
